Separate local and remote model methods

merlinc opened the pull request:
- Refactor code to ES6 classes
- Separate model.js into:
  - local-model - simple non callback methods
  - remote-model - methods using callbacks and/or request
  - util - non-model specific utilities

HughePaul commented on lib/local-model.js:
```diff
@@ -0,0 +1,105 @@
+        let old = this.toJSON(),
+            changed = {};
+
+        _.each(attrs, function (value, key) {
```
es6ify inline functions

HughePaul commented on lib/local-model.js:
```diff
@@ -0,0 +1,105 @@
+        if (!options.silent && !_.isEmpty(changed)) {
+            _.each(changed, function (value, key) {
+                this.emit('change:' + key, this.get(key), old[key]);
+            }, this);
```
es6ify inline functions, and remove explicit context

HughePaul commented on lib/remote-model.js:
```diff
@@ -0,0 +1,186 @@
+        } else if (!options) {
+            options = {};
+        }
+        this.prepare(function (err, data) {
```
es6ify inline functions, and remove bind

HughePaul commented on lib/remote-model.js:
```diff
@@ -0,0 +1,186 @@
+        }
+        this.prepare(function (err, data) {
+
+            if (err) { return callback(err); }
```
I think inline single line returns look better without braces

HughePaul commented on lib/util.js:
```diff
@@ -0,0 +1,17 @@
+'use strict';
```
new line after strict

HughePaul commented on lib/util.js:
```diff
@@ -0,0 +1,17 @@
+    }
+}
+
+module.exports = new Util();
```
This doesn't doesn't need to be a class, this would be a collection of functions.
urlKeys should be a constant. It can be a constant value of this files returned object and still tested.

HughePaul commented on package.json:
```diff
@@ -4,11 +4,13 @@
     "lint": "eslint ./",
     "unit": "mocha test/spec/ --recursive --require test/helper",
-    "cover": "istanbul cover _mocha -- test/spec/ --recursive --require test/helper",
-    "check-coverage": "istanbul check-coverage --statement 100 --branch 100 --function 100 --line 100",
```
needs `.istanbul` file

HughePaul commented on test/spec/spec.remote-model.js:
```diff
@@ -1,14 +1,22 @@
         model = new Model();
+
         cb = sinon.stub();
         apiRequest = {};
```
declare before instantly executed returns line.

HughePaul commented on test/spec/spec.remote-model.js:
```diff
@@ -29,228 +37,210 @@ describe('Model model', function () {
+
+    describe('request', () => {
+        let settings,
             bodyData;
```
I prefer lets on one line or each with a let

HughePaul commented on test/spec/spec.remote-model.js:
```diff
@@ -29,228 +37,210 @@ describe('Model model', function () {
             expect(options.body).to.not.be.ok;
         });
 
         it('can parse failiure when no callback given', function (done) {
```
es6ify functions

HughePaul commented on test/spec/spec.remote-model.js:
```diff
@@ -29,228 +37,210 @@ describe('Model model', function () {
+        beforeEach(() => {
             model.set('name', 'Test name');
-            model.url = function () { return 'http://example.com:3002/foo/bar'; };
+            model.url = () => { return 'http://example.com:3002/foo/bar'; };
```
braces and return not required

HughePaul commented on test/spec/spec.remote-model.js:
```diff
@@ -651,64 +641,64 @@ describe('Model model', function () {
         it('calls callback with an error if API response returns an error code', function (done) {
-            Model._request.yieldsAsync(null, fail);
+            requestStub.yieldsAsync(null, fail);
             model.delete(function (e, data, responseTime) {
```
es6ify functions

merlinc commented:
I've addressed all of your review comments except for the ones about `util.js`. The subsequent refactoring removes the need for this entire class.

Code coverage checking has also been restored, but it has dropped below 100%.

> Statements   : 96.1% ( 148/154 )
Branches     : 93.14% ( 95/102 )
Functions    : 95.24% ( 20/21 )
Lines        : 96.08% ( 147/153 )

In this case this should not considered a problem as the subsequent refactoring restores 100% test coverage.
